### ISSUE TITLE
metaMDS() fails when computing wascores if supplied a long expression for `comm`

### ISSUE THREAD

**Issue by gavinsimpson** [vegandevs/vegan#133]
Here's a reproducible example:

```r
library("vegan")
data(varespec)
set.seed(1)
metaMDS(subset(varespec, select = colSums(varespec) > 0, subset = rowSums(varespec) > 0))
```

which produces

```r
> metaMDS(subset(varespec, select = colSums(varespec) > 0, subset = rowSums(varespec) > 0))
Square root transformation
Wisconsin double standardization
Run 0 stress 0.1843196 
Run 1 stress 0.2455912 
Run 2 stress 0.2169407 
Run 3 stress 0.2313231 
Run 4 stress 0.1974406 
Run 5 stress 0.1858402 
Run 6 stress 0.1948414 
Run 7 stress 0.2265717 
Run 8 stress 0.222507 
Run 9 stress 0.2023215 
Run 10 stress 0.2673177 
Run 11 stress 0.1976151 
Run 12 stress 0.1852397 
Run 13 stress 0.2341085 
Run 14 stress 0.1955867 
Run 15 stress 0.2137409 
Run 16 stress 0.2109638 
Run 17 stress 0.1825658 
... New best solution
... procrustes: rmse 0.04169825  max resid 0.1521436 
Run 18 stress 0.1843197 
Run 19 stress 0.2570119 
Run 20 stress 0.3760596 
Error in parse(text = attr(dis, "commname")) : 
  <text>:1:94: unexpected ')'
1: wisconsin(sqrt(subset(varespec, select = colSums(varespec) > 0, subset = rowSums(varespec) > )
```

The problem occurs [here](https://github.com/vegandevs/vegan/blob/master/R/metaMDS.R#L50) as we try to recover the possibly transformed data from the calling frame:

```r
    if (wascores) {
        ## transformed data
        comm <- eval.parent(parse(text=attr(dis, "commname")))
        wa <- wascores(points, comm, expand = expand)
    }
```

The failure is due to the `deparse(substitute(comm))` step early on in `metaMDS()` breaking the expression up into more than a single string:

```r
foo <- function(comm) {
    deparse(substitute(comm))
}
foo(subset(varespec, select = colSums(varespec) > 0, subset = rowSums(varespec) > 0))
```

which produces:

```r
> foo(subset(varespec, select = colSums(varespec) > 0, subset = rowSums(varespec) > 0))
[1] "subset(varespec, select = colSums(varespec) > 0, subset = rowSums(varespec) > "
[2] "    0)"
```

A simple solution would just be to concatenate the elements of the character vector returned by the `deparse(substitute(comm))` line, using `paste0()` (or equivalent). But it seems wrong to be deparsing and parsing R expressions like this. Ideally we would capture the `comm` argument unevaluated and then evaluate it again later but that is going to be tricky as `metaMDSdist()` modifies this expression (in character form).

However, I'm wondering why we don't just return the community data as a component of `metaMDSdist()`'s returned object?

Thoughts @jarioksa ?

**Comment by jarioksa** [comment]
Minimal solution is to use wider strings when deparsing the name:
```diff
diff --git a/R/metaMDS.R b/R/metaMDS.R
index fafa804..672ceba 100644
--- a/R/metaMDS.R
+++ b/R/metaMDS.R
@@ -6,7 +6,7 @@
               plot = FALSE, previous.best,  ...) 
 {
     engine <- match.arg(engine)
-    commname <- deparse(substitute(comm))
+    commname <- deparse(substitute(comm), width.cutoff = 500L)
     ## metaMDS was written for community data which should be all
     ## positive. Check this here, and set arguments so that they are
     ## suitable for non-negative data.
```
Internally evaluating `comm` and using that in `wascores()` could be a  more robust solution, though. Somebody should only have a look at that. The minimal fix above would solve this particular problem, and if we have no time to try the more fundamental fix, we may go with this.

**Comment by jarioksa** [comment]
There seems to be need of some rethinking of design. Why we have this `commname`  is that input `comm`  can be either dissimilarities (`"dist"` object, symmetric square matrix) or a community matrix. If it is a community matrix, we call `metaMDSdist` that returns dissimilarities. All other information is returned as `attribute`s of the `"dist"` object. We could add the edited `comm` data as an extra attribute. As `metaMDSdist` manipulates `comm`  it also updates the `commname` to match the manipulations, and this edited call will be one `attribute` of dissimilarities. Although `commname` is documented in `metaMDSdist`, users cannot set it in their `metaMDS` call.

**Comment by gavinsimpson** [comment]
I would prefer that we just modify `metaMDSdist()` to return the modified community data as an attribute. There doesn't seem to be much point in evaluating an expression to create an object we already could have at our disposal. I have a fix I testing.

**Comment by jarioksa** [comment]
I agree with you: returning an `attribute` is the real fix. However, I think we *also* need the `deparse` length fix because we print the edited data command.

**Comment by gavinsimpson** [comment]
:+1: I'm doing that too; but I want to see how complex the code would be to ignore the `width.cutoff` entirely and put the string back together after it has been butched. Someone will inevitably find a way to stick more than 500 bytes in there :-)

**Comment by jarioksa** [comment]
Closed with PR #135
